doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,215 @@
+            object is called (the object has to be loaded again into the cpu cache, cache miss for 
+            every member which is not used by the method).
+
+    **Adjustments:**
+      - Create the class with a template bool array size argument to be more flexible and memory efficient.
+
+    **Note:**
+      - Sadly, we cannot pursue this approach in the WaitSet since it is event 
+        and state based at the same time and this approach supports only event 
+        driven triggering.
+
+  - add class `ConditionVariableWithOriginTracingSignaler` (I know, better name)
+  
+    **Reason:** it behaves quiete differently then the `ConditionVariable`. The `ConditionVariable` tells you:
+    someone has signalled you and you have to find out who it was. The `ConditionVariableWithOriginTracing`
+    would tell you: A, B and C have signalled you.
+
+    ```cpp
+    class ConditionVariableWithOriginTracingSignaler {
+      public:
+        // id = corresponds to the id in the ConditionVariable atomic_bool array to 
+        // identify the trigger origin easily
+        void notifyOne(const uint64_t id); 
+    };
+    ```
+
+  - add class `ConditionVariableWithOriginTracingWaiter`
+  ```cpp
+  class ConditionVariableWithOriginTracingWaiter {
+    public:
+      // returns a list of ids which have notified the conditionVariable
+      cxx::vector<uint64_t> wait(); 
+      // sets the atomic bool to false in m_triggeredBy[id]
+      void resetWaitState(const uint64_t id) noexcept; 
+  };
+  ```
+
+We maybe introduce some abstraction for this. One thought is to introduce a 
+`SignalVector` where you can acquire a `Notifyier` which then signals the 
+`ConditionVariable` and with the correct id.
+
+#### ReactAl 
+```cpp
+enum class ReactAlErrors {
+  CALLBACK_CAPACITY_EXCEEDED
+};
+
+template<uint64_t CallbackCapacity>
+class ReactAl {
+  public:
+    template<typename EventOrigin, typename EventType>
+    cxx::expected<ReactAlErrors> attachEvent(
+      EventOrigin & origin, const EventType & eventType,
+      const cxx::function_ref<void(EventOrigin&)> & callback
+    ) noexcept;
+
+    template<typename EventOrigin, typename EventType>
+    void detachEvent( EventOrigin & origin, const EventType & eventType );
+};
+
+```
+ 
+  - Contains a thread which will wake up and iterate through ConditionVariable
+    bool array. If an entry is true, set it to false and then call the 
+    corresponding callback.
+
+#### Class which is attachable to ReactAl 
+
+```cpp
+class SomeSubscriber {
+  //...
+  // index is the index to the bool inside of the condition variable, maybe it 
+  // is later hidden inside of some abstraction
+  //
+  // OutOfScopeCallback_t - callback which should be called when SomeSubscriber
+  //                        goes out of scope and has to deregister itself from 
+  //                        the reactal
+  void attachCallback(SubscriberEvent event, ConditionVariable & cv, ConditionVariable::index_t & index
```

Comment:
Update example?

Reply:
Will be done in the final task when this design document is updated after the implementation is finished. At the moment it should provide you with the idea on how it could look like.